traffic_ops/traffic_ops_golang/user/user_test.go

```diff
@@ -0,0 +1,347 @@
+package user
+
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import (
+	"testing"
+	"time"
+
+	"github.com/apache/trafficcontrol/lib/go-tc"
+	"github.com/apache/trafficcontrol/lib/go-util"
+)
+
+var (
+	user = &tc.UserV4{
+		AddressLine1:      util.Ptr("line1"),
+		AddressLine2:      util.Ptr("line2"),
+		ChangeLogCount:    8,
+		City:              util.Ptr("city"),
+		Company:           util.Ptr("company"),
+		Country:           util.Ptr("country"),
+		Email:             util.Ptr("test@email.com"),
+		FullName:          util.Ptr("Testy Mctestface"),
+		GID:               nil,
+		ID:                util.Ptr(1),
+		LastAuthenticated: util.Ptr(time.Now()),
+		LastUpdated:       time.Now(),
+		LocalPassword:     util.Ptr("pw"),
+		NewUser:           false,
+		PhoneNumber:       util.Ptr("999-999-9999"),
+		PostalCode:        util.Ptr("11111-1111"),
+		PublicSSHKey:      nil,
+		RegistrationSent:  util.Ptr(time.Now()),
+		Role:              "role",
+		StateOrProvince:   util.Ptr("state"),
+		Tenant:            nil,
+		TenantID:          0,
+		Token:             nil,
+		UCDN:              "",
+		UID:               nil,
+		Username:          "testy",
+	}
+	oldUser = defineUser()
+)
+
+func defineUser() tc.User {
+	u := tc.User{
+		Username:             util.Ptr(user.Username),
+		RegistrationSent:     nil,
+		LocalPassword:        user.LocalPassword,
+		ConfirmLocalPassword: user.LocalPassword,
+		RoleName:             util.Ptr(user.Role),
+	}
+	u.AddressLine1 = user.AddressLine1
+	u.AddressLine2 = user.AddressLine2
+	u.City = user.City
+	u.Company = user.Company
+	u.Country = user.Country
+	u.Email = user.Email
+	u.FullName = user.FullName
+	u.GID = user.GID
+	u.ID = user.ID
+	u.LastUpdated = tc.TimeNoModFromTime(user.LastUpdated)
+	u.NewUser = util.Ptr(user.NewUser)
+	u.PhoneNumber = user.PhoneNumber
+	u.PostalCode = user.PostalCode
+	u.PublicSSHKey = user.PublicSSHKey
+	u.Role = nil
+	u.StateOrProvince = user.StateOrProvince
+	u.Tenant = user.Tenant
+	u.TenantID = util.Ptr(user.TenantID)
+	u.Token = user.Token
+	u.UID = user.UID
+	return u
+}
+
+func TestDowngrade(t *testing.T) {
+	old := user.Downgrade()
+
+	if *old.FullName != *oldUser.FullName {
+		t.Fatalf("expected FullName to be equal, got %v and %v", *old.FullName, *oldUser.FullName)
```

Comment:
nit: `%v` seems like a bad idea when the type is known. You get more safety out of being more specific.